### PR TITLE
added a few questions

### PR DESCRIPTION
**Pull request by dong77**

### CODE REVIEW

File: src/EmergencyMultisig.sol [hunk 1]
```diff
@@ -59,6 +59,7 @@ contract EmergencyMultisig is IEmergencyMultisig, IMembership, PluginUUPSUpgrade
     struct MultisigSettings {
         bool onlyListed;
         uint16 minApprovals;
```
Comment (brickpop) [inline src/EmergencyMultisig.sol:61]: The standard multisig is the source of truth regarding the address list, whereas the emergency multisig uses the standard multisig's list. This is to save storage and transactions to maintain, assuming that both had the same members.

File: src/Multisig.sol [hunk 2]
```diff
@@ -225,6 +225,8 @@ contract Multisig is IMultisig, IMembership, PluginUUPSUpgradeable, ProposalUpgr
         // Create the proposal
         Proposal storage proposal_ = proposals[proposalId];
         proposal_.metadataURI = _metadataURI;
+
+        // Q(danielw): shall we validate `_destinationPlugin`?
```
Comment (brickpop) [inline src/Multisig.sol:229]: Permission checks are made on the destination plugin, rather than at the source. 

The `createProposal()` method on `OptimisticTokenVotingPlugin` has a modifier `auth(PROPOSER_PERMISSION_ID)`, which is only granted to the two multisig's. 

Regarding the case where `_destinationPlugin` was a random address, there wouldn't be an inventive to do so because the multisig's have no permission anywhere else.

File: src/Multisig.sol [hunk 3]
```diff
@@ -233,6 +235,8 @@ contract Multisig is IMultisig, IMembership, PluginUUPSUpgradeable, ProposalUpgr
 
         for (uint256 i; i < _destinationActions.length;) {
             proposal_.destinationActions.push(_destinationActions[i]);
+
+            // Q(danielw): With the new solidity version, using `++i` in the for loop is good enough.
```
Comment (brickpop) [inline src/Multisig.sol:239]: This change applies for compiler versions equal or greater than 0.8.22
Unfortunately, the dependencies we use force us to stay on version 0.8.17 for now, where such optimization was not present

File: src/Multisig.sol [hunk 4]
```diff
@@ -415,5 +419,6 @@ contract Multisig is IMultisig, IMembership, PluginUUPSUpgradeable, ProposalUpgr
     /// @dev This empty reserved space is put in place to allow future versions to add new
     /// variables without shifting down storage in the inheritance chain.
     /// https://docs.openzeppelin.com/contracts/4.x/upgradeable#storage_gaps
+    /// Q(danielw): will `multisigSettings` and `lastMultisigSettingsChange` share the same slot?
```
Comment (brickpop) [inline src/Multisig.sol:422]: As far as I can see, there should be 3 storage slots, each being:
- `mapping(uint256 => Proposal) internal proposals`
- `MultisigSettings public multisigSettings` where `bool + uint16 and uint64` should only consume one slot
- `uint64 public lastMultisigSettingsChange`

Did I miss something?

File: src/OptimisticTokenVotingPlugin.sol [hunk 5]
```diff
@@ -62,6 +62,7 @@ contract OptimisticTokenVotingPlugin is
         bytes metadataURI;
         IDAO.Action[] actions;
         uint256 allowFailureMap;
+        // Q(danielw): I believe the code to upgrade `aggregatedL2Balance` is missing.
```
Comment (brickpop) [inline src/OptimisticTokenVotingPlugin.sol:65]: Correct, in this version our goal was to lay all the foundation to be future compatible with L2 voting, but leaving the final L2 aggregation logic for the next stage, as we will have clarity on this.

Should this be built for the initial release?

File: src/OptimisticTokenVotingPlugin.sol [hunk 6]
```diff
@@ -184,13 +185,18 @@ contract OptimisticTokenVotingPlugin is
     }
 
     /// @inheritdoc IOptimisticTokenVoting
+
     function totalVotingPower(uint256 _timestamp) public view returns (uint256) {
         return votingToken.getPastTotalSupply(_timestamp);
+        // Q(danielw): we should probably need to remove votes of `taikoL1` which holds a lot of TAIKO as bond deposits.
```
Comment (brickpop) [inline src/OptimisticTokenVotingPlugin.sol:191]: If this balance is unrelated to L2 being up or not (effectiveVotingPower) then sure, let's update the expression.

File: src/OptimisticTokenVotingPlugin.sol [hunk 7]
```diff
@@ -184,13 +185,18 @@ contract OptimisticTokenVotingPlugin is
...
+        // return votingToken.getPastTotalSupply(_timestamp) - votingToken.getPastVotes(address(taikoL1), _timestamp);
     }
 
     /// @inheritdoc IOptimisticTokenVoting
     function bridgedVotingPower(uint256 _timestamp) public view returns (uint256) {
+         // Q(danielw): I think we should use `taikoERC20Vault`, not `taikoBridge`.
```
Comment (brickpop) [inline src/OptimisticTokenVotingPlugin.sol:197]: My bad. I had been using that value from the list of addresses you provided, assuming that it was like an alias. Should replacing the name of the variable globally be enough, or is there any use for the `taikoBridge` related to the DAO?

File: src/OptimisticTokenVotingPlugin.sol [hunk 8]
```diff
@@ -203,6 +209,8 @@ contract OptimisticTokenVotingPlugin is
     }
 
     /// @notice Determines whether L2 votes are currently usable for voting
+    // Q(daniel): not sure if we should use the last L2 block or the last verified L2 block...
```
Comment (brickpop) [inline src/OptimisticTokenVotingPlugin.sol:212]: If we should, then we will :)

File: src/OptimisticTokenVotingPlugin.sol [hunk 9]
```diff
@@ -257,6 +270,7 @@ contract OptimisticTokenVotingPlugin is
         }
 
         // The bridge cannot vote directly. It must use a dedicated function.
+        // Q(danielw): better to exclude taikoL1 and taikoERC20Vault as well.
```
Comment (brickpop) [inline src/OptimisticTokenVotingPlugin.sol:273]: Ok 👍 
Question: Can `taikoBridge` be removed in general?

File: src/OptimisticTokenVotingPlugin.sol [hunk 10]
```diff
@@ -345,6 +359,7 @@ contract OptimisticTokenVotingPlugin is
             snapshotTimestamp = block.timestamp - 1; // The snapshot timestamp must in the past to protect the transaction against backrunning transactions causing census changes.
         }
 
+        // Q(danielw): taikoBridge => taikoERC20Vault
```
Comment (brickpop) [inline src/OptimisticTokenVotingPlugin.sol:362]: 👍

File: src/OptimisticTokenVotingPlugin.sol [hunk 11]
```diff
@@ -542,6 +557,7 @@ contract OptimisticTokenVotingPlugin is
 
         // No more L2 vetoes can be registered
         // return false if no L2 votes have been aggregated until now
+        // Q(danielw): what if the aggregated L2 votes is actually zero? Maybe use another boolean var to track?
```
Comment (brickpop) [inline src/OptimisticTokenVotingPlugin.sol:560]: Not sure I follow
You mean using a new proposal variable, which is already precomputed following a similar logic while removing this function?

File: src/PublicKeyRegistry.sol [hunk 12]
```diff
@@ -17,6 +17,8 @@ contract PublicKeyRegistry {
     /// @notice Raised when the public key of the given user has already been set
     error AlreadySet();
 
+
+    /// Q(danielw): it seems if an incorrect pubkey is set, then there is no way to correct it.
```
Comment (brickpop) [inline src/PublicKeyRegistry.sol:21]: Correct. The rationale behind this being:
- The derived wallet generation is an automated process, with no human intervention
- One wallet can only generate one derived wallet, given the same signing payload
- If someone registers an incorrect public key, that would be very, very strange. This would mean that proposals could be read (potentially) by someone else, other than the signer.

If such an edge case arises, my recommendation would be to make it obvious to the rest of Security Council members, when the bad actor sees that he/she needs to start using another wallet and asks the multisig to replace the address.

The UI only encrypts proposal's data for public keys registered by multisig signers, but without adding ZK proofs, there's no way to verify that the public key of each signer was generated using the private key of their wallet. 

We can bypass this check and allow anyone to change their public key, but updating keys might become a non-feature that opens the door to members silently sharing data with someone else, without needing to expose their personal wallet.

File: src/OptimisticTokenVotingPlugin.sol [hunk 13]
```diff
@@ -219,6 +227,11 @@ contract OptimisticTokenVotingPlugin is
         if ((_block.proposedAt + governanceSettings.l2InactivityPeriod) < block.timestamp) return false;
 
         return true;
+
+
+        // new impl suggested by Daniel W
+        // (,, uint64 lastBlockVerifiedTimestamp) = taikoL1.getLastVerifiedBlock();
+        // return _block.proposedAt + governanceSettings.l2InactivityPeriod >= block.timestamp;
```
Comment (brickpop) [inline src/OptimisticTokenVotingPlugin.sol:234]: It appears as  `_block.proposedAt` should also be removed?

Would this implementation look correct to you?

```solidity
        // Actively disabled L2 voting?
        if (governanceSettings.skipL2) return false;
        // Is the L1 bridge paused?
        else if (taikoL1.paused()) return false;

        uint64 _id = taikoL1.slotB().numBlocks;
        // No L2 blocks yet
        if (_id == 0) return false;

        (,, uint64 lastBlockVerifiedTimestamp) = taikoL1.getLastVerifiedBlock();
        return (lastBlockVerifiedTimestamp + governanceSettings.l2InactivityPeriod) >= block.timestamp;
```

Comment (brickpop) [reply src/OptimisticTokenVotingPlugin.sol:234]: We should also upgrade to use version 1.8.0 of the protocol. Is this correct?

File: src/EmergencyMultisig.sol (hunk 1 above)
Comment (dong77) [reply src/EmergencyMultisig.sol:61]: I don't follow. But I can ask later in our next call.

File: src/OptimisticTokenVotingPlugin.sol (hunk 5 above)
Comment (dong77) [reply src/OptimisticTokenVotingPlugin.sol:65]: Nope, just making sure I understand it right.

File: src/OptimisticTokenVotingPlugin.sol (hunk 6 above)
Comment (dong77) [reply src/OptimisticTokenVotingPlugin.sol:191]: The balance of TaikoL1 is unrelated to the state of L2.

File: src/OptimisticTokenVotingPlugin.sol (hunk 7 above)
Comment (dong77) [reply src/OptimisticTokenVotingPlugin.sol:197]: The bridge contract is not supposed to hold any TAIKO token, so you can ignore its token balance. the bridge contract will only be used here to aggregate L2 votes later.

File: src/OptimisticTokenVotingPlugin.sol (hunk 13 above)
Comment (dong77) [reply src/OptimisticTokenVotingPlugin.sol:234]: Yes please, then later we will release a 1.9.0 for our Ontake hardfork.

File: src/OptimisticTokenVotingPlugin.sol (hunk 9 above)
Comment (dong77) [reply src/OptimisticTokenVotingPlugin.sol:273]: To calculate L1 votes, we can ignore `taikoBridge`, but it will be used to aggregate vetos from L2.

File: src/OptimisticTokenVotingPlugin.sol (hunk 11 above)
Comment (dong77) [reply src/OptimisticTokenVotingPlugin.sol:560]: can `proposal_.aggregatedL2Balance == 0` mean EITHER the actual L2 veto is zero OR somehow the actual L2 veto was not bridged/aggregated to L1 for this proposal (for example, L2 is down)? If this is the case, shall we differentiate these two cases?

File: src/PublicKeyRegistry.sol (hunk 12 above)
Comment (dong77) [reply src/PublicKeyRegistry.sol:21]: Works for me.

File: src/OptimisticTokenVotingPlugin.sol (hunk 8 above)
Comment (dong77) [reply src/OptimisticTokenVotingPlugin.sol:212]: I think we should, but this can be done later. I still need to merge my PR first.

File: src/EmergencyMultisig.sol (hunk 1 above)
Comment (brickpop) [reply src/EmergencyMultisig.sol:61]: What I mean is that instead of having two separate lists that need to be maintained separately, we only have one and the standard multisig acts as the database. 

This is assuming that both plugins are controlled by the same group of people. If this wasn't the case, then we would need to undo this change and use two almost identical multisigs, but with independent settings/signers.

File: src/OptimisticTokenVotingPlugin.sol (hunk 11 above)
Comment (brickpop) [reply src/OptimisticTokenVotingPlugin.sol:560]: Oh I see. This edge case was briefly commented on the chat without a clear outcome. Since this doesn't affect the future-proofness of the plugin it wasn't brought back again. 

Intuitively, if there are 0 aggregated L2 votes and the L2 is currently down, we should only use the L1 census. But I wonder if this could lead to unforeseen race conditions (someone knowing that there is an upcoming L2 batch that somehow can't make it to the L1, and getting to execute with a favorable L1 census)

File: src/OptimisticTokenVotingPlugin.sol (hunk 8 above)
Comment (brickpop) [reply src/OptimisticTokenVotingPlugin.sol:212]: Which PR, what changes do you mean?

File: src/OptimisticTokenVotingPlugin.sol (hunk 13 above)
Comment (brickpop) [reply src/OptimisticTokenVotingPlugin.sol:234]: I see that the current signature on 1.8.0 is `function getLastVerifiedBlock() external view returns (uint64 blockId_, bytes32 blockHash_, bytes32 stateRoot_)`

Whereas the example code is retrieving `(,, uint64 lastBlockVerifiedTimestamp)`, which doesn't seem to match.
Is this the change that will be landing on 1.9.0?

File: src/EmergencyMultisig.sol (hunk 1 above)
Comment (dantaik) [reply src/EmergencyMultisig.sol:61]: Ah, I see, so `addresslistSource` in EmergencyMultisig.sig points to the Multisig contract which manages the actual list of approvers, right?

File: src/OptimisticTokenVotingPlugin.sol (hunk 8 above)
Comment (dantaik) [reply src/OptimisticTokenVotingPlugin.sol:212]: https://github.com/taikoxyz/taiko-mono/pull/17868 This one, which is part of the main branch, not in a release branch.

File: src/OptimisticTokenVotingPlugin.sol (hunk 13 above)
Comment (dantaik) [reply src/OptimisticTokenVotingPlugin.sol:234]: Yes, it will be part of 1.9.0

File: src/OptimisticTokenVotingPlugin.sol (hunk 11 above)
Comment (dantaik) [reply src/OptimisticTokenVotingPlugin.sol:560]: I think we can address this later before we enable L2 voting. May I trouble you to add a comment in the code?

File: src/EmergencyMultisig.sol (hunk 1 above)
Comment (brickpop) [reply src/EmergencyMultisig.sol:61]: @dantaik correct
We can improve the naming if this feels confusing, but it basically points to the address list to use

File: src/OptimisticTokenVotingPlugin.sol (hunk 11 above)
Comment (brickpop) [reply src/OptimisticTokenVotingPlugin.sol:560]: @dantaik Sure, let's add a comment for this

I am currently addressing the review's feedback on this separate branch https://github.com/aragon/taiko-contracts/tree/f/review-feedback as your changes are living on a separate repo

File: src/OptimisticTokenVotingPlugin.sol (hunk 13 above)
Comment (dantaik) [reply src/OptimisticTokenVotingPlugin.sol:234]: See https://github.com/aragon/taiko-contracts/pull/39
